language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -1,50 +1,163 @@
-    TyLam () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyApp _ ty ty') = TyApp () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM (TyFun _ ty ty') = TyFun () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM x@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
-    cloneSt <- get
-    case IM.lookup (unUnique u) cloneSt of
-        Just n  -> pure (TyVar () n)
-        Nothing -> pure (void x)
-cloneTypeM x@TyBuiltin{}    = pure (void x)
-cloneTypeM x@TyInt{}        = pure (void x)
+
+import           Control.Applicative
+import           Control.Monad.Reader
+import           Control.Monad.State
+
+import           Data.Foldable
+import           Data.Functor.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.IntSet               as IS
+import qualified Data.List.NonEmpty        as NE
+import           Data.Maybe                (fromMaybe)
+import           Lens.Micro
+
+type Cloning m = (MonadState CloneSt m, MonadReader CloneWhat m, MonadQuote m)
+
+data CloneSt = CloneSt {
+    csMappingStack :: NE.NonEmpty (IM.IntMap Unique) -- ^ A scoped stack of unique mappings.
+    , csBound      :: IS.IntSet -- ^ The set of bound variables seen so far.
+    }
+
+emptyState :: CloneSt
+emptyState = CloneSt {csMappingStack=mempty NE.:| [], csBound=mempty}
+
+addMapping :: Cloning m => Int -> Unique -> m ()
+addMapping oldU newU = modify $ \(CloneSt (top NE.:| rest) bound) -> CloneSt (IM.insert oldU newU top NE.:| rest) bound
+
+lookupMapping :: Cloning m => Unique -> m (Maybe Unique)
+lookupMapping (Unique u) = do
+    mappings <- gets csMappingStack
+    pure $ foldl' (<|>) Nothing (fmap (IM.lookup u) mappings)
+
+-- | Run the given action with an additional scope popped on the stack, and then pop it off afterwards.
+withScope :: Cloning m => m a -> m a
+withScope act = do
+    modify $ \(CloneSt mappings bound) -> CloneSt (mempty NE.<| mappings) bound
+    a <- act
+    modify $ \(CloneSt (_ NE.:| rest) bound) -> CloneSt (fromMaybe (mempty NE.:| []) (NE.nonEmpty rest)) bound
+    pure a
+
+addBound :: Cloning m => Unique -> m ()
+addBound u@(Unique i) = do
+    modify $ \(CloneSt mappings bound) -> CloneSt mappings (IS.insert i bound)
+    -- if we are cloning bound variables then we should make sure that we make a fresh
+    -- mapping for this variable, even if it already appears in the mappings, since there
+    -- might be name shadowing
+    doClone <- shouldClone u
+    when doClone (void $ freshenUnique u)
+
+isBound :: Cloning m => Unique -> m Bool
+isBound (Unique u) = IS.member u <$> gets csBound
+
+shouldClone :: Cloning m => Unique -> m Bool
+shouldClone u = do
+    bound <- isBound u
+    what <- ask
+    pure $ if bound then cloneBound what else cloneFree what
+
+data CloneWhat = Bound | Free | Both
+
+cloneBound :: CloneWhat -> Bool
+cloneBound = \case
+    Bound -> True
+    Free -> False
+    Both -> True
+
+cloneFree :: CloneWhat -> Bool
+cloneFree = \case
+    Bound -> False
+    Free -> True
```

Comment:
Why do we have a case for renaming `Free` which is not used anywhere?

Reply:
Mostly because it was the third logical option and I didn't see any reason not to include it. I can remove it until we need it, but I'm not sure it makes much difference.

Reply:
I prefer not having dead code around, so I'd prefer not including it until we need it.